README.md
[![Ruby](https://github.com/whotwagner/domain_name_format_validator/actions/workflows/main.yml/badge.svg)](https://github.com/whotwagner/domain_name_format_validator/actions/workflows/main.yml)
[![Gem Version](https://badge.fury.io/rb/domain_name_format_validator.svg)](https://badge.fury.io/rb/domain_name_format_validator)

domain_name_format_validator
============================

Ever needed to validate the format of a domain name? This gem will validate any domain name
represented in ASCII. 

The scope of this gem is deliberately focused on validating the format of domain names. It
simply answers the question: "Is this a real domain name?" Using this command,
you can make a realistic assessment about whether you want to store a domain
name or URL in your database. This gem will tell you:
1. that a domain is or is not valid 
2. if it's not valid, what the errors are. 

Some existing gems for domain name validation use insecure regular expressions.
For example if you have the following regex-pattern: `^do_some_checks$`, the
check can be bypassed by domainn names like: `example.com\n<script>alert('xss')</script>`. 
Such a bypass doesn't work with this gem.

_Please note that this gem is a fork of https://github.com/dkeener/domain_name_validator which seems to be abandoned_

How It Works
------------

To validate a domain name:

```ruby
    require 'domain_name_format_validator'

    if DomainNameFormatValidator.valid?("example.com")
      # Do something
    end
```

What about error messages? If a domain isn't valid, it's often desirable to
find out why the domain ewasn't valid. To do this, simply pass an array into
the "validate" message as the optional second argument.

```ruby
    require 'domain_name_format_validator'

    errs = DomainNameFormatValidator.errors("example.123")
    unless errs.empty?
      puts("Errors: #{errs.inspect}")
    end
```

This generates the following output:

    Errors: ["The top-level domain (the extension) cannot be numerical"]

This gem should make it easy to validate domain names.

About Domain Names
------------------

Domain names provide a unique, memorizable name to represent numerically
addressable Internet resources. They also provide a level of abstraction that
allows the underlying Internet address to be changed while still referencing
a resource by its domain name. The domain name space is managed by the 
Internet Corporation for Assigned Names and Numbers (ICANN).

The right-most label of a domain name is referred to as the top-level domain,
or TLD. A limited set of top-level domain names, and two-character country
codes, have been standardized. The Internet Assigned Numbers Authority (IANA)
maintains an annotated list of top-level domains, as well as a list of
"special use," or reserved, top-level domain names.

* http://www.iana.org/domains/root/db/
* http://www.iana.org/assignments/special-use-domain-names/special-use-domain-names.xml

Domain names follow some very detailed rules:

* The maximum length of a domain name is 253 characters.

* A domain name is divided into "labels" separated by periods. The maximum
  number of labels is 127.

* The maximum length of any label within a domain name is 63 characters.

* No label, including TLDs, can begin or end with a dash.

* Top-level domain names cannot be all numeric.

* The right-most label must be either a recognized TLD or a 2-letter country
  code. The only exception is for international domain names

* Top-level domain names cannot be all numeric.

* Domain names may not begin with a period.

* The characters allowed in labels are a subset of the ASCII character set, consisting of 
  characters a through z, A through Z, digits 0 through 9, and hyphen.


Internationalized Domain Names
------------------------------

What about internationalized domain names? ICANN approved the Internationalized
Domain Name (IDNA) system in 2003. This standard allows for Unicode domain
names to be encoded into ASCII using Punycode. Essentially, a label may contain
"xn--" as a prefix, followed by the Punycode representation of a Unicode string,
resulting in domain names such as xn--kbenhavn-54.eu. Note that there are also
some approved Unicode TLDs.

The process of rendering an internationalized domain name in ASCII via 
Punycode is called normalization. This gem will validate a normalized domain
name, but not a Unicode domain name. Note, however, that it currently does not
validate normalized TLDs against ICANN's list of valid TLDs.

It's also unclear whether the "xn--" prefix should count against the label
size limit of 63 characters. In the absence of specific guidelines, and because
I've never actually seen an overly long label, I have chosen to apply the limit
irregardless of the presence of the "xn--" prefix within a label.

Requirements
------------

This is a Ruby gem with no run-time dependencies on anything else. It has
been tested under Ruby 3.0.0p0 but might work with older Ruby versions too.

Install
-------

Installation doesn't get much simpler than this:

    gem install domain_name_format_validator


Author
------

This gem was refactored and relaunched by [Wolfgang Hotwagner](https://github.com/whotwagner/domain_name_format_validator).
The original code was written by [David Keener](https://github.com/dkeener/domain_name_validator).  


Contributors
------------

Many thanks for the support of General Dynamics and the Department of
Homeland Security (DHS). And more specifically on input from Andrew Finch,
Josh Lentz, Jonathan Quigg and Dave Roberts.

YOUR SUPPORT
------------

Every contribution is welcome. Please feel free to open Pull-Requests.


ChangeLog.md
1.0 (2021-08-18)
----------------

* Fixed issues with newline in domain-names

* Changed behavior for Top-Level-Domain-checks: they are now limited by 63
  characters.

* Changed class into a ruby module and added a error-function as requested in
  https://github.com/dkeener/domain_name_validator/issues/6

* Added tests for common top level domains and some random valid domains

* Added tests for regex bypass

* Ruby style: added rubocop and fixed styling issues

* Replaced rspec with minitest

* New name 'domain_name_format_validator' indicates that the format will be
  validated and not if it is an existing domain or top-level-domain

* Modified Rakefile it can run test-tasks and rubycop-tasks

0.5 (2013-10-24)
----------------

* By request, the gem now ignores leading/trailing whitespace when validating
  domain names. This includes ignoring newlines at the end of domain names.

* Remove timestamp from gemspec so Rubygems will calculate gem dates for us.

* (0.5.1 - 2013-11-06) The validate method now avoids accidental memoization
  on the part of the caller by ensuring that the contents of the array
  argument have been cleared. Previously, if the caller called the method
  multiple times using the same array, then all calls would result in FALSE
  after the first FALSE was legitemately encountered. Yes, somebody did
  this.

0.4 (2013-07-17)
----------------

* By request, added rudimentary TLD checking. A TLD, the right-most label of a
  domain name, should be either 2 or 3 characters, unless it's "aero", "arpa",
  "museum" or begins with "xn--" (for a normalized Unicode TLD). This check
  does not validate against a master list of TLD's, but basically just fails
  a domain name if the TLD basically could not be valid under any
  circumstances. This validation would, for example, reject a domain name
  such as "test.domain".

* (0.4.1 - 2013-07-17) Added "info" as a valid extra-long TLD. How I missed
  this one, I have no idea.

* (0.4.2 - 2013-07-17) Added a check for zero-length domain names, either
  empty strings or Ruby nil values being passed in. In practice, these edge
  cases were actually happening.

* (0.4.3 - 2013-08-01) For the TLD reality check, IANA (the Internet Assigned
  Numbers Authority) recognizes some long TLD's that were not on the original
  list for this gem, such as ".travel", ".jobs", etc. To reiterate, the gem
  does not validate against the full list of TLD's or effective TLD's, but 
  it does validate that the TLD length COULD be a valid TLD. As an example,
  a TLD of 14 characters would be rejected because there are no valid domains
  that could match that pattern.

* (0.4.3 - 2013-08-01) Added the "license" attribute to the gemspec. This is
  in response to an issue that was opened on GitHub.

* (0.4.4 - 2013-08-28) Fixed bug involving proper handling of domain names with
  capital letters. Reported by a user.

0.3 (2013-06-28)
----------------

* Added a check, because labels cannot begin with a period.
* Updated documentation, plus defined Road Map for future changes.

0.2 (2013-06-17)
----------------

First version built and deployed as a gem on a real project. Released on
Rubygems.org for the first time.

* Added more RSpec tests for improved test coverage.
* Enhanced documentation.


0.1 (2013-06-10)
----------------

Initial working version of the code, with minimal RSpec tests. Released on
GitHub for review, but not yet published as a gem to the Ruby community.


spec/spec_helper.rb
require File.expand_path(
    File.join(File.dirname(__FILE__), %w[.. lib domain_name_format_validator]))
        
RSpec.configure do |config|
  # == Mock Framework
  #
  # RSpec uses it's own mocking framework by default. If you prefer to
  # use mocha, flexmock or RR, uncomment the appropriate line:
  #
  # config.mock_with :mocha
  # config.mock_with :flexmock
  # config.mock_with :rr
end


test/test_helper.rb
require 'domain_name_format_validator'


lib/domain_name_format_validator/settings.rb
# frozen_string_literal: true

# Some constants and error-codes for DomainNameFormatValidator
# are defined here
module DomainNameFormatValidator
  MAX_DOMAIN_LENGTH = 253
  MAX_LABEL_LENGTH = 63
  MAX_LEVELS = 127
  MAX_TLD_LENGTH = 63
  MIN_LEVELS = 2
  MIN_TLD_LENGTH = 2

  ERRS = {
    bogus_tld: "Malformed TLD: Could not possibly match any valid TLD",
    illegal_chars: "Domain label contains an illegal character",
    illegal_start: "No domain name may start with a period",
    label_dash_begin: "No domain label may begin with a dash",
    label_dash_end: "No domain label may end with a dash",
    max_domain_size: "Maximum domain length of 253 exceeded",
    max_label_size: "Maximum domain label length of 63 exceeded",
    max_level_size: "Maximum domain level limit of 127 exceeded",
    min_level_size: "Minimum domain level limit of 2 not achieved",
    top_numerical: "The top-level domain (TLD) cannot be numerical",
    top_illegal_chars: "The top-level domain (TLD) must only contain a-z 0-9 and -",
    zero_size: "Zero-length domain name",
    not_a_string: "Domain must be a string"
  }.freeze
end


lib/domain_name_format_validator/validator.rb
# frozen_string_literal: true

# The purpose of this module is to provide a simple capability for validating
# domain names represented in ASCII.
module DomainNameFormatValidator
  # Validates the proper formatting of a normalized domain name, i.e. - a
  # domain that is represented in ASCII. Thus, international domain names are
  # supported and validated, if they have undergone the required IDN
  # conversion to ASCII. The validation rules are:
  #
  # 1. The maximum length of a domain name is 253 characters.
  # 2. A domain name is divided into "labels" separated by periods. The maximum
  #    number of labels (including the top-level domain as a label) is 127.
  # 3. The maximum length of any label within a domain name is 63 characters.
  # 4. No label, including top-level domains, can begin or end with a dash.
  # 5. Top-level names cannot be all numeric.
  # 6. A domain name cannot begin with a period.

  # this internal function validates a single label
  # and is used by validate_parts?
  def self.validate_part?(part, errs = [])
    errs << ERRS[:max_label_size] if part.size > MAX_LABEL_LENGTH
    errs << ERRS[:label_dash_begin] if part[0] == "-"
    errs << ERRS[:label_dash_end] if part[-1] == "-"
    errs << ERRS[:illegal_chars] unless part.match(/\A[a-z0-9\-\_]+\Z/)
  end

  # This internal function validates the labels of a domain name
  def self.validate_parts?(parts, errs = [])
    errs << ERRS[:max_level_size] if parts.size > MAX_LEVELS
    errs << ERRS[:min_level_size] if parts.size < MIN_LEVELS
    errs << ERRS[:illegal_start] if parts.first[0] == "."
    parts.each do |part|
      validate_part?(part, errs)
    end
    errs
  end

  # this internal function validates the top level domain
  # if its nummerical only, if illegal characters occur
  # or if the length of the tld is not valid
  def self.validate_tld?(tld, errs = [])
    errs << ERRS[:top_numerical] if tld.match(/\A[0-9]+\Z/)
    errs << ERRS[:top_illegal_chars] unless tld.match(/\A[a-z0-9\-]+\Z/)
    errs << ERRS[:bogus_tld] if tld.size < MIN_TLD_LENGTH || tld.size > MAX_TLD_LENGTH
    errs
  end

  # This internal function validates the domain if its nil or zero
  def self.validate_args?(domain, errs = [])
    if domain.nil?
      errs << ERRS[:zero_size]
    else
      return [ERRS[:not_a_string]] unless domain.is_a? String

      domain = domain.strip
      errs << ERRS[:zero_size] if domain.size.zero?
    end
    errs
  end

  # This function validates domain names and returns an array
  # with errors or an empty array if no error occurred.
  # see: https://github.com/dkeener/domain_name_validator/issues/6
  def self.errors(domain)
    errs = []
    errs = validate_args?(domain, errs)
    if errs.size.zero?
      errs << ERRS[:max_domain_size] if domain.size > MAX_DOMAIN_LENGTH
      parts = domain.downcase.split(".")
      errs = validate_parts?(parts, errs)
      errs = validate_tld?(parts.last, errs)
    end
    errs
  end

  # This function validates domain names and returns true or false
  def self.valid?(domain)
    errs = errors(domain)
    errs.size.zero?   # TRUE if valid, FALSE otherwise
  end
end
